R-package/tests/testthat/test_basic.R

```diff
@@ -2042,3 +2042,147 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
   expect_equal(pred1, pred2)
 
 })
+
+context("monotone constraints")
+
+.generate_trainset_for_monotone_constraints_tests <- function(x3_to_categorical) {
+  n_samples <- 3000L
+  x1_positively_correlated_with_y <- rnorm(n = n_samples)
+  x2_negatively_correlated_with_y <- rnorm(n = n_samples)
+  x3_negatively_correlated_with_y <- rnorm(n = n_samples)
+  if (x3_to_categorical) {
+    x3_negatively_correlated_with_y <- as.integer(abs(runif(n_samples) / 0.25))
+    categorical_features <- 3L
+  } else {
+    x3_negatively_correlated_with_y <- runif(n_samples)
```

Comment:
Why does this line needed?

Reply:
oh! It isn't, you're that I have duplicate definitions. The one a few lines up can be removed.

```r
x3_negatively_correlated_with_y <- rnorm(n = n_samples)
```

Reply:
fixed in https://github.com/microsoft/LightGBM/commit/e7b2048a6f8b9e42e5dda4b9012d8c9dcc06a9d7